{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Table of Contents\n",
    " <p><div class=\"lev1\"><a href=\"#Task-1.-Compiling-Ebola-Data\"><span class=\"toc-item-num\">Task 1.&nbsp;&nbsp;</span>Compiling Ebola Data</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-2.-RNA-Sequences\"><span class=\"toc-item-num\">Task 2.&nbsp;&nbsp;</span>RNA Sequences</a></div>\n",
    " <div class=\"lev1\"><a href=\"#Task-3.-Class-War-in-Titanic\"><span class=\"toc-item-num\">Task 3.&nbsp;&nbsp;</span>Class War in Titanic</a></div></p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "DATA_FOLDER = 'Data' # Use the data folder provided in Tutorial 02 - Intro to Pandas.\n",
    "%matplotlib inline\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set_context('notebook')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 1. Compiling Ebola Data\n",
    "\n",
    "The `DATA_FOLDER/ebola` folder contains summarized reports of Ebola cases from three countries (Guinea, Liberia and Sierra Leone) during the recent outbreak of the disease in West Africa. For each country, there are daily reports that contain various information about the outbreak in several cities in each country.\n",
    "\n",
    "Use pandas to import these data files into a single `Dataframe`.\n",
    "Using this `DataFrame`, calculate for *each country*, the *daily average per month* of *new cases* and *deaths*.\n",
    "Make sure you handle all the different expressions for *new cases* and *deaths* that are used in the reports."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution\n",
    "---\n",
    "##### Assumption : For this exercise, we assume that when there is no value in a field, it does not mean 0, it means that we don't know the value. Thus, for the daily average per month, we do this average using existing data.\n",
    "\n",
    "First of all, we need to read all of the data which is in the csv format. So for this, we need to retrieve all csv file names for some country. \n",
    "Thus using a ls command seems to be the best option here. After that, we retrieve and concatenate the csv files.\n",
    "We then filter the rows by their description, looking for the new cases and deaths related descriptions. \n",
    "\n",
    "Right after, we rename these descriptions to be either 'New cases' or 'Deaths'.\n",
    "There may be multiple 'New cases' or 'Deaths' possibilities for each date, however, as we will see, we deal with them using groupby operations.\n",
    "\n",
    "We then recompute the total using the results per county, as sometimes the total is not defined even if some county have some values. We assign the result to the total field when it's not defined. And for rows where there are less than 3 values defined, which are the Date, the Descripton and the total (or some county values that are used to recompute the total), we drop them.\n",
    "\n",
    "After that we can group_by the month, the day and the description, to obtain at most one result per (day, 'New cases'), and at most one result per (day, 'Deaths'). \n",
    "\n",
    "After that, we can do the daily average per month, by doing a groupby over the first and third index columns (month and description), and a mean of each bucket. We rename the indexes and the column.\n",
    "We do this procedure for each country, with different column names, 'New cases' and 'Deaths' descriptions, and when we have the three DataFrames, we concatenate them.\n",
    "\n",
    "However, it's not that easy for the whole data : in fact for Liberia, in the month of december the data was entered incorrectly. The data given in the Description fields 'new case/s (suspected)', 'new case/s (probable)' and 'new case/s (confirmed)' is cumulative, compared to before when it was the result of the day. Fortunately, as for this month we have continuous days : from day 1 to day 9 without gap, we can infer the missing data by using the cumulative cases and deaths fields in Description : 'cumulative confirmed, probable and suspected cases' and 'total death/s in confirmed, \\n probable, suspected cases'. By using the 2014-11-30 day as a starting point, we can create the New cases and Deaths data we wanted. We do it in the country_DF_cumulative function.\n",
    "\n",
    "Because of this, we need to specify to the country_DF_mass function (working with non-cumulative fields) that liberia will not need to work with all csv files, as the december subpart is already used in the other method. The lsnames field is there for that.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Question : in the case\n",
    "\n",
    "* 0 \t2014-10-01 \tNew cases of suspects \t28\n",
    "* 1 \t2014-10-01 \tNew cases of probables \t0\n",
    "* 2 \t2014-10-01 \tNew cases of confirmed \t6\n",
    "* 3 \t2014-10-01 \tTotal new cases registered so far \t34\n",
    "* 15 \t2014-10-01 \tNew cases of confirmed among health workers \t2\n",
    "\n",
    "Do we count the New cases of confirmed among health workers as part of New cases of confirmed, so that the total is correct, or is it a new category ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Daily average per month</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Month</th>\n",
       "      <th>Description</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">Guinea</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">8</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>3.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>25.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>3.562500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>19.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>15.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>34.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"14\" valign=\"top\">Liberia</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">6</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>5.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">7</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>4.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>8.545455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">8</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>23.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>37.222222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>37.608696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>63.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>28.040000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>45.560000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">11</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>15.538462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>30.266667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">12</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>14.777778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>16.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"10\" valign=\"top\">Sierra Leone</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">8</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>5.550000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>38.350000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">9</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>4.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>71.620690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">10</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>4.952381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>124.214286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">11</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>172.578947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">12</th>\n",
       "      <th>Deaths</th>\n",
       "      <td>5.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>New cases</th>\n",
       "      <td>226.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Daily average per month\n",
       "             Month Description                         \n",
       "Guinea       8     Deaths                      3.400000\n",
       "                   New cases                  25.800000\n",
       "             9     Deaths                      3.562500\n",
       "                   New cases                  19.625000\n",
       "             10    Deaths                     15.000000\n",
       "                   New cases                  34.000000\n",
       "Liberia      6     Deaths                      2.000000\n",
       "                   New cases                   5.714286\n",
       "             7     Deaths                      4.272727\n",
       "                   New cases                   8.545455\n",
       "             8     Deaths                     23.222222\n",
       "                   New cases                  37.222222\n",
       "             9     Deaths                     37.608696\n",
       "                   New cases                  63.833333\n",
       "             10    Deaths                     28.040000\n",
       "                   New cases                  45.560000\n",
       "             11    Deaths                     15.538462\n",
       "                   New cases                  30.266667\n",
       "             12    Deaths                     14.777778\n",
       "                   New cases                  16.000000\n",
       "Sierra Leone 8     Deaths                      5.550000\n",
       "                   New cases                  38.350000\n",
       "             9     Deaths                      4.571429\n",
       "                   New cases                  71.620690\n",
       "             10    Deaths                      4.952381\n",
       "                   New cases                 124.214286\n",
       "             11    Deaths                      2.000000\n",
       "                   New cases                 172.578947\n",
       "             12    Deaths                      5.500000\n",
       "                   New cases                 226.333333"
      ]
     },
     "execution_count": 296,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_filenames(path):\n",
    "    filenames = !ls $path\n",
    "    return filenames\n",
    "\n",
    "def get_cols(df, to_remove):\n",
    "    cols = pd.Series(df.columns)\n",
    "    return cols[~cols.isin(to_remove)]\n",
    "\n",
    "def country_DF_mass(country, date_c, description_c, total_c, filters_new_cases, filters_deaths, lsnames=''):\n",
    "    \n",
    "    PATH = ''\n",
    "    \n",
    "    if(not lsnames):\n",
    "        PATH = DATA_FOLDER + '/ebola/' + country + '_data/'\n",
    "        country_files = get_filenames(PATH)\n",
    "    else:\n",
    "        country_files = lsnames\n",
    "\n",
    "    country_DF = pd.concat([pd.read_csv(PATH+filename, parse_dates=[date_c], na_values=['-']) \n",
    "                           for filename in country_files]) \n",
    "\n",
    "    country_DF = country_DF.loc[[ description.lower() in filters_new_cases or\n",
    "                                    description.lower() in filters_deaths\n",
    "                                    for description in country_DF[description_c].values]]\n",
    "    \n",
    "    country_DF.loc[[description.lower() in filters_new_cases for description in country_DF[description_c]], description_c] = 'New cases'\n",
    "    country_DF.loc[[description.lower() in filters_deaths for description in country_DF[description_c]], description_c] = 'Deaths'\n",
    "    \n",
    "    cols = get_cols(country_DF, to_remove=[date_c, description_c, total_c])\n",
    "    \n",
    "    country_DF = country_DF.dropna(thresh = 3) #drop every entry containing no data, as we can't infer anything (it's not 0)\n",
    "    recomputed_total = country_DF[cols].apply(pd.to_numeric).sum(axis=1)\n",
    "\n",
    "    country_DF[total_c] = pd.to_numeric(country_DF[total_c]).combine_first(recomputed_total)\n",
    "    \n",
    "    country_computation = pd.DataFrame(country_DF.groupby([country_DF[date_c].dt.month, country_DF[date_c].dt.day, country_DF[description_c]])[total_c].sum())\n",
    "    country_computation = country_computation.groupby(level=[0,2]).mean()\n",
    "    country_computation.index.names = ['Month', 'Description']\n",
    "    country_computation.columns = ['Daily average per month']\n",
    "    return country_computation\n",
    "\n",
    "\n",
    "def country_DF_cumulative(country, date_c, description_c, total_c, filters_new_cases, filters_deaths, date_begin= '2014-11-30', date_end='2014-12-09'):\n",
    "    \"\"\"To use this function, the data must be in following days, => no gap\"\"\"\n",
    "    PATH = DATA_FOLDER + '/ebola/' + country + '_data/'\n",
    "    \n",
    "    country_DF = pd.concat([pd.read_csv(PATH + date_begin + '.csv', parse_dates=[date_c], na_values=['-']), \n",
    "                            pd.read_csv(PATH + date_end + '.csv', parse_dates=[date_c], na_values=['-'])]) \n",
    "    \n",
    "    country_DF = country_DF.loc[[ description.lower() in filters_new_cases or\n",
    "                                    description.lower() in filters_deaths\n",
    "                                    for description in country_DF[description_c].values]]\n",
    "    \n",
    "    cols = get_cols(country_DF, to_remove=[date_c, description_c, total_c])\n",
    "    recomputed_total = country_DF[cols].apply(pd.to_numeric).sum(axis=1)\n",
    "    country_DF[total_c] = pd.to_numeric(country_DF[total_c]).combine_first(recomputed_total)\n",
    "    \n",
    "    country_DF.loc[[description.lower() in filters_new_cases for description in country_DF[description_c]], description_c] = 'New cases'\n",
    "    country_DF.loc[[description.lower() in filters_deaths for description in country_DF[description_c]], description_c] = 'Deaths'\n",
    "    \n",
    "    days_between = (pd.to_datetime(date_end) - pd.to_datetime(date_begin)).days\n",
    "    \n",
    "    country_DF = country_DF.groupby(country_DF[description_c])[total_c].apply(lambda x: x.iloc[1] - x.iloc[0])/days_between\n",
    "    country_DF = pd.DataFrame(pd.concat({pd.to_datetime(date_end).month: country_DF}))\n",
    "    country_DF.index.names = ['Month', 'Description']\n",
    "    country_DF.columns = ['Daily average per month']\n",
    "    \n",
    "    return country_DF\n",
    "\n",
    "\n",
    "guinea = country_DF_mass('guinea', 'Date', 'Description', 'Totals', ['total new cases registered so far'], ['new deaths registered', 'new deaths registered today'])\n",
    "\n",
    "liberia_mass_files = !ls 'Data/ebola/liberia_data'/2014-[0.1][^2]*\n",
    "liberia = country_DF_mass('liberia', 'Date', 'Variable', 'National', ['new case/s (suspected)', 'new case/s (probable)', 'new case/s (confirmed)'], ['newly reported deaths'], liberia_mass_files)\n",
    "liberia_december = country_DF_cumulative('liberia', 'Date', 'Variable', 'National', ['cumulative confirmed, probable and suspected cases'], ['total death/s in confirmed, \\n probable, suspected cases'])\n",
    "liberia = pd.concat([liberia, liberia_december])\n",
    "\n",
    "sl = country_DF_mass('sl', 'date', 'variable', 'National', ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed'], ['etc_new_deaths'])\n",
    "\n",
    "pd.concat({'Guinea': guinea, 'Liberia': liberia, 'Sierra Leone': sl})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "def country_DF(country, date_c, description_c, total_c, filters_new_cases, filters_deaths):\n",
    "    path = DATA_FOLDER + '/ebola/' + country + '_data/'\n",
    "    country_files = !ls $path\n",
    "\n",
    "    country_DF = pd.concat([pd.read_csv(path+filename, usecols=[date_c, description_c, total_c], parse_dates=[date_c], index_col=False) \n",
    "                           for filename in country_files]) \n",
    "    \n",
    "\n",
    "    country_DF = country_DF.loc[[ description.lower() in filters_new_cases or\n",
    "                                    description.lower() in filters_deaths\n",
    "                                    for description in country_DF[description_c].values]]\n",
    "\n",
    "    country_DF.loc[[description.lower() in filters_new_cases for description in country_DF[description_c]], description_c] = 'New cases'\n",
    "    country_DF.loc[[description.lower() in filters_deaths for description in country_DF[description_c]], description_c] = 'Deaths'\n",
    "    country_DF[total_c] = pd.to_numeric(country_DF[total_c])\n",
    "    \n",
    "\n",
    "    country_computation = pd.DataFrame(country_DF.groupby([country_DF[date_c].dt.month, country_DF[date_c].dt.day, country_DF[description_c]])[total_c].sum())\n",
    "    country_computation = country_computation.groupby(level=[0,2]).mean()\n",
    "    country_computation.index.names = ['Month', 'Description']\n",
    "    country_computation.columns = ['Daily average per month']\n",
    "    return country_computation\n",
    "\n",
    "guinea = country_DF('guinea', 'Date', 'Description', 'Totals', ['total new cases registered so far'], ['new deaths registered', 'new deaths registered today'])\n",
    "liberia = country_DF('liberia', 'Date', 'Variable', 'National', ['new case/s (suspected)', 'new case/s (probable)', 'new case/s (confirmed)'], ['newly reported deaths'])\n",
    "sl = country_DF('sl', 'date', 'variable', 'National', ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed'], ['death_suspected', 'death_probable', 'death_confirmed'])\n",
    "\n",
    "pd.concat({'Guinea': guinea, 'Liberia': liberia, 'Sierra Leone': sl})\n",
    "\"\"\"\n",
    "pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "def country_DF(country, date_c, description_c, total_c, filters_new_cases, filters_deaths):\n",
    "    path = DATA_FOLDER + '/ebola/' + country + '_data/'\n",
    "    country_files = !ls $path\n",
    "\n",
    "    country_DF = pd.concat([pd.read_csv(path+filename, parse_dates=[date_c], na_values=['-']) \n",
    "                           for filename in country_files]) \n",
    "\n",
    "    country_DF = country_DF.loc[[ description.lower() in filters_new_cases or\n",
    "                                    description.lower() in filters_deaths\n",
    "                                    for description in country_DF[description_c].values]]\n",
    "    \n",
    "    cols = pd.Series(country_DF.columns)\n",
    "    cols = cols[~cols.isin([date_c, description_c, total_c])]\n",
    "    \n",
    "    #print(country_DF.dropna(thresh = 3))\n",
    "    #print('-'*200)\n",
    "    country_DF = country_DF.dropna(thresh = 3) #drop every entry containing no data, as we can't infer anything (it's not 0)\n",
    "    recomputed_total = country_DF[cols].apply(pd.to_numeric).sum(axis=1)\n",
    "    \n",
    "    #tmp = pd.DataFrame()\n",
    "    #tmp[\"recomputed\"] = recomputed_total\n",
    "    #tmp[\"total\"] = pd.to_numeric(country_DF[total_c])\n",
    "    #tmp[\"combined\"] = pd.to_numeric(country_DF[total_c]).combine_first(recomputed_total)\n",
    "    #print(tmp)\n",
    "    \n",
    "    country_DF[total_c] = pd.to_numeric(country_DF[total_c]).combine_first(recomputed_total)\n",
    "    #print(country_DF[[date_c, description_c, total_c]])\n",
    "\n",
    "    country_DF.loc[[description.lower() in filters_new_cases for description in country_DF[description_c]], description_c] = 'New cases'\n",
    "    country_DF.loc[[description.lower() in filters_deaths for description in country_DF[description_c]], description_c] = 'Deaths'\n",
    "    #country_DF[total_c] = pd.to_numeric(country_DF[total_c])\n",
    "    \n",
    "    \n",
    "    country_computation = pd.DataFrame(country_DF.groupby([country_DF[date_c].dt.month, country_DF[date_c].dt.day, country_DF[description_c]])[total_c].sum())\n",
    "    country_computation = country_computation.groupby(level=[0,2]).mean()\n",
    "    country_computation.index.names = ['Month', 'Description']\n",
    "    country_computation.columns = ['Daily average per month']\n",
    "    return country_computation\n",
    "\n",
    "guinea = country_DF('guinea', 'Date', 'Description', 'Totals', ['total new cases registered so far'], ['new deaths registered', 'new deaths registered today'])\n",
    "liberia = country_DF('liberia', 'Date', 'Variable', 'National', ['new case/s (suspected)', 'new case/s (probable)', 'new case/s (confirmed)'], ['newly reported deaths'])\n",
    "sl = country_DF('sl', 'date', 'variable', 'National', ['new_noncase', 'new_suspected', 'new_probable', 'new_confirmed'], ['etc_new_deaths'])\n",
    "\n",
    "pd.concat({'Guinea': guinea, 'Liberia': liberia, 'Sierra Leone': sl})\n",
    "\"\"\"\n",
    "pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 2. RNA Sequences\n",
    "\n",
    "In the `DATA_FOLDER/microbiome` subdirectory, there are 9 spreadsheets of microbiome data that was acquired from high-throughput RNA sequencing procedures, along with a 10<sup>th</sup> file that describes the content of each. \n",
    "\n",
    "Use pandas to import the first 9 spreadsheets into a single `DataFrame`.\n",
    "Then, add the metadata information from the 10<sup>th</sup> spreadsheet as columns in the combined `DataFrame`.\n",
    "Make sure that the final `DataFrame` has a unique index and all the `NaN` values have been replaced by the tag `unknown`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution\n",
    "---\n",
    "First of all, I query the microbiome file names, using ls. I remove the extension part.\n",
    "Then, I read the xls files and convert it to a DataFrame, and I put an additional column (BARCODE), corresponding to the file name, with the same value for the whole DataFrame.\n",
    "I concatenate those DataFrames into one and I put the index (Phylum) as a column.\n",
    "I can now do an inner join between the big DataFrame and the metadata DataFrame, on the BARCODE column. I set the index ('Phylum', 'BARCODE'), as it's unique. And I can replace all occurences of NaN by 'unknown', as requested in the statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Counts</th>\n",
       "      <th>GROUP</th>\n",
       "      <th>SAMPLE</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phylum</th>\n",
       "      <th>BARCODE</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Desulfurococcaceae Ignisphaera</th>\n",
       "      <th>MID1</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Desulfurococcales Pyrodictiaceae Pyrolobus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Sulfolobales Sulfolobaceae Stygiolobus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Crenarchaeota\" Thermoprotei Thermoproteales Thermofilaceae Thermofilum</th>\n",
       "      <th>MID1</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanocellales Methanocellaceae Methanocella</th>\n",
       "      <th>MID1</th>\n",
       "      <td>7</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methanosarcinaceae Methanimicrococcus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" \"Methanomicrobia\" Methanosarcinales Methermicoccaceae Methermicoccus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Ferroglobus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglobales Archaeoglobaceae Geoglobus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloplanus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Haloquadratum</th>\n",
       "      <th>MID1</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Halosimplex</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronobacterium</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Halobacteria Halobacteriales Halobacteriaceae Natronomonas</th>\n",
       "      <th>MID1</th>\n",
       "      <td>4</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanococci Methanococcales Methanocaldococcaceae Methanotorris</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Methanopyri Methanopyrales Methanopyraceae Methanopyrus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>12</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea \"Euryarchaeota\" Thermoplasmata Thermoplasmatales Thermoplasmatales_incertae_sedis Thermogymnomonas</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Acidothermaceae Acidothermus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Beutenbergiaceae Salana</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Bogoriellaceae Bogoriella</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cellulomonadaceae Tropheryma</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Corynebacteriaceae Turicella</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Cryptosporangiaceae Cryptosporangium</th>\n",
       "      <th>MID1</th>\n",
       "      <td>53</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Dermabacter</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermabacteraceae Devriesea</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Dermatophilaceae Kineosphaera</th>\n",
       "      <th>MID1</th>\n",
       "      <td>3</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Frankineae_incertae_sedis Fodinicola</th>\n",
       "      <th>MID1</th>\n",
       "      <td>1</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Blastococcus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>99</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Geodermatophilaceae Geodermatophilus</th>\n",
       "      <th>MID1</th>\n",
       "      <td>17</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Actinobacteria\" Actinobacteria Actinomycetales Intrasporangiaceae Kribbia</th>\n",
       "      <th>MID1</th>\n",
       "      <td>2</td>\n",
       "      <td>EXTRACTION CONTROL</td>\n",
       "      <td>unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillaceae Neptuniibacter</th>\n",
       "      <th>MID9</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oceanospirillales_incertae_sedis Spongiispira</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Oceanospirillales Oleiphilaceae Oleiphilus</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pasteurellales Pasteurellaceae Haemophilus</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Alkanindiges</th>\n",
       "      <th>MID9</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Moraxellaceae Perlucidibaca</th>\n",
       "      <th>MID9</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azomonas</th>\n",
       "      <th>MID9</th>\n",
       "      <td>805</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azorhizophilus</th>\n",
       "      <th>MID9</th>\n",
       "      <td>862</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Azotobacter</th>\n",
       "      <th>MID9</th>\n",
       "      <td>12</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Pseudomonadales Pseudomonadaceae Pseudomonas</th>\n",
       "      <th>MID9</th>\n",
       "      <td>3534</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Cycloclasticus</th>\n",
       "      <th>MID9</th>\n",
       "      <td>4</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Piscirickettsiaceae Sulfurivirga</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichaceae Beggiatoa</th>\n",
       "      <th>MID9</th>\n",
       "      <td>25</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Thiotrichales Thiotrichales_incertae_sedis Fangia</th>\n",
       "      <th>MID9</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Alkanibacter</th>\n",
       "      <th>MID9</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Sinobacteraceae Sinobacter</th>\n",
       "      <th>MID9</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Aquimonas</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Frateuria</th>\n",
       "      <th>MID9</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Ignatzschineria</th>\n",
       "      <th>MID9</th>\n",
       "      <td>2</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Proteobacteria\" Gammaproteobacteria Xanthomonadales Xanthomonadaceae Luteibacter</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Synergistetes\" Synergistia Synergistales Synergistaceae Aminomonas</th>\n",
       "      <th>MID9</th>\n",
       "      <td>3</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Anaeroplasmatales Anaeroplasmataceae Asteroleplasma</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Tenericutes\" Mollicutes Haloplasmatales Haloplasmataceae Haloplasma</th>\n",
       "      <th>MID9</th>\n",
       "      <td>11</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Caldimicrobium</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermodesulfobacteria\" Thermodesulfobacteria Thermodesulfobacteriales Thermodesulfobacteriaceae Thermodesulfatator</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Geotoga</th>\n",
       "      <th>MID9</th>\n",
       "      <td>43</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Thermotogae\" Thermotogae Thermotogales Thermotogaceae Kosmotoga</th>\n",
       "      <th>MID9</th>\n",
       "      <td>16</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria \"Verrucomicrobia\" Opitutae Opitutales Opitutaceae Opitutus</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Chlorarachniophyceae</th>\n",
       "      <th>MID9</th>\n",
       "      <td>1</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria Cyanobacteria Cyanobacteria  Chloroplast Streptophyta</th>\n",
       "      <th>MID9</th>\n",
       "      <td>10</td>\n",
       "      <td>Control 2</td>\n",
       "      <td>stool</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2396 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            Counts  \\\n",
       "Phylum                                             BARCODE           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1          7   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1          2   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal... MID1          3   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot... MID1          3   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1          7   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1          1   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1          4   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1          2   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1          4   \n",
       "Archaea \"Euryarchaeota\" Methanococci Methanococ... MID1          1   \n",
       "Archaea \"Euryarchaeota\" Methanopyri Methanopyra... MID1         12   \n",
       "Archaea \"Euryarchaeota\" Thermoplasmata Thermopl... MID1          2   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          2   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          1   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          1   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          1   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          2   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1         53   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          1   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          2   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          3   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          1   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1         99   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1         17   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1          2   \n",
       "...                                                            ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9         10   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9          1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9          1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9          1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9         12   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9          4   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9        805   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9        862   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9         12   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       3534   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9          4   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9          1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9         25   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9          2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          3   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          1   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          3   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9          1   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista... MID9          3   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat... MID9          1   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal... MID9         11   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9          1   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9          1   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9         43   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9         16   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ... MID9          1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9          1   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9         10   \n",
       "\n",
       "                                                                         GROUP  \\\n",
       "Phylum                                             BARCODE                       \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Methanococci Methanococ... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Methanopyri Methanopyra... MID1     EXTRACTION CONTROL   \n",
       "Archaea \"Euryarchaeota\" Thermoplasmata Thermopl... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     EXTRACTION CONTROL   \n",
       "...                                                                        ...   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9              Control 2   \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista... MID9              Control 2   \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat... MID9              Control 2   \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal... MID9              Control 2   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9              Control 2   \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9              Control 2   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9              Control 2   \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9              Control 2   \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ... MID9              Control 2   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9              Control 2   \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9              Control 2   \n",
       "\n",
       "                                                             SAMPLE  \n",
       "Phylum                                             BARCODE           \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1     unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Desulfuroc... MID1     unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Sulfolobal... MID1     unknown  \n",
       "Archaea \"Crenarchaeota\" Thermoprotei Thermoprot... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" \"Methanomicrobia\" Metha... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Archaeoglobi Archaeoglo... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Halobacteria Halobacter... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Methanococci Methanococ... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Methanopyri Methanopyra... MID1     unknown  \n",
       "Archaea \"Euryarchaeota\" Thermoplasmata Thermopl... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "Bacteria \"Actinobacteria\" Actinobacteria Actino... MID1     unknown  \n",
       "...                                                             ...  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria O... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria P... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria T... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Proteobacteria\" Gammaproteobacteria X... MID9       stool  \n",
       "Bacteria \"Synergistetes\" Synergistia Synergista... MID9       stool  \n",
       "Bacteria \"Tenericutes\" Mollicutes Anaeroplasmat... MID9       stool  \n",
       "Bacteria \"Tenericutes\" Mollicutes Haloplasmatal... MID9       stool  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9       stool  \n",
       "Bacteria \"Thermodesulfobacteria\" Thermodesulfob... MID9       stool  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9       stool  \n",
       "Bacteria \"Thermotogae\" Thermotogae Thermotogale... MID9       stool  \n",
       "Bacteria \"Verrucomicrobia\" Opitutae Opitutales ... MID9       stool  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9       stool  \n",
       "Bacteria Cyanobacteria Cyanobacteria  Chloropla... MID9       stool  \n",
       "\n",
       "[2396 rows x 3 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PATH = DATA_FOLDER + '/microbiome/'\n",
    "files = !cd $PATH; ls MI*\n",
    "filenames = map(lambda file: file[:file.find('.')], files)\n",
    "\n",
    "rna = pd.concat([pd.read_excel(PATH + filename + '.xls', sheetname=0, header=None, names= [\"Counts\"], index_col=0)\n",
    "                 .assign(BARCODE=filename) for filename in filenames])\n",
    "rna['Phylum'] = rna.index.values\n",
    "\n",
    "metadata = pd.read_excel(DATA_FOLDER + '/microbiome/metadata.xls')\n",
    "\n",
    "merged = pd.merge(rna, metadata, on='BARCODE').set_index(['Phylum', 'BARCODE'])\n",
    "merged = merged.replace(np.nan, 'unknown')\n",
    "\n",
    "#print(merged.index.is_unique) #prints True\n",
    "#print(merged.isnull().values.any()) #prints False\n",
    "merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Task 3. Class War in Titanic\n",
    "\n",
    "Use pandas to import the data file `Data/titanic.xls`. It contains data on all the passengers that travelled on the Titanic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<hr><h2>Data frame:titanic3</h2>1309 observations and 14 variables, maximum # NAs:1188<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Name</td><td>Labels</td><td>Units</td><td>Levels</td><td>Storage</td><td>NAs</td></tr>\n",
       "<tr><td>pclass</td><td></td><td></td><td><a href=\"#pclass\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>survived</td><td>Survived</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>name</td><td>Name</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>sex</td><td></td><td></td><td><a href=\"#sex\"><div align=right>  2</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>age</td><td>Age</td><td>Year</td><td><div align=right></div></td><td>double</td><td><div align=right> 263</div></td></tr>\n",
       "<tr><td>sibsp</td><td>Number of Siblings/Spouses Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>parch</td><td>Number of Parents/Children Aboard</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>ticket</td><td>Ticket Number</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>fare</td><td>Passenger Fare</td><td>British Pound (\\243)</td><td><div align=right></div></td><td>double</td><td><div align=right>   1</div></td></tr>\n",
       "<tr><td>cabin</td><td></td><td></td><td><a href=\"#cabin\"><div align=right>187</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>embarked</td><td></td><td></td><td><a href=\"#embarked\"><div align=right>  3</div></a></td><td>integer</td><td><div align=right>   2</div></td></tr>\n",
       "<tr><td>boat</td><td></td><td></td><td><a href=\"#boat\"><div align=right> 28</div></a></td><td>integer</td><td><div align=right>   0</div></td></tr>\n",
       "<tr><td>body</td><td>Body Identification Number</td><td></td><td><div align=right></div></td><td>double</td><td><div align=right>1188</div></td></tr>\n",
       "<tr><td>home.dest</td><td>Home/Destination</td><td></td><td><div align=right></div></td><td>character</td><td><div align=right>   0</div></td></tr>\n",
       "</TABLE>\n",
       "<hr>\n",
       "<TABLE BORDER>\n",
       "<tr><td>Variable</td><td>Levels</td></tr>\n",
       "<tr><td><a name=\"pclass\">pclass</a></td><td>1st</td></tr>\n",
       "<tr><td></td><td>2nd</td></tr>\n",
       "<tr><td></td><td>3rd</td></tr>\n",
       "<tr><td><a name=\"sex\">sex</a></td><td>female</td></tr>\n",
       "<tr><td></td><td>male</td></tr>\n",
       "<tr><td><a name=\"cabin\">cabin</a></td><td></td></tr>\n",
       "<tr><td></td><td>A10</td></tr>\n",
       "<tr><td></td><td>A11</td></tr>\n",
       "<tr><td></td><td>A14</td></tr>\n",
       "<tr><td></td><td>A16</td></tr>\n",
       "<tr><td></td><td>A18</td></tr>\n",
       "<tr><td></td><td>A19</td></tr>\n",
       "<tr><td></td><td>A20</td></tr>\n",
       "<tr><td></td><td>A21</td></tr>\n",
       "<tr><td></td><td>A23</td></tr>\n",
       "<tr><td></td><td>A24</td></tr>\n",
       "<tr><td></td><td>A26</td></tr>\n",
       "<tr><td></td><td>A29</td></tr>\n",
       "<tr><td></td><td>A31</td></tr>\n",
       "<tr><td></td><td>A32</td></tr>\n",
       "<tr><td></td><td>A34</td></tr>\n",
       "<tr><td></td><td>A36</td></tr>\n",
       "<tr><td></td><td>A5</td></tr>\n",
       "<tr><td></td><td>A6</td></tr>\n",
       "<tr><td></td><td>A7</td></tr>\n",
       "<tr><td></td><td>A9</td></tr>\n",
       "<tr><td></td><td>B10</td></tr>\n",
       "<tr><td></td><td>B101</td></tr>\n",
       "<tr><td></td><td>B102</td></tr>\n",
       "<tr><td></td><td>B11</td></tr>\n",
       "<tr><td></td><td>B18</td></tr>\n",
       "<tr><td></td><td>B19</td></tr>\n",
       "<tr><td></td><td>B20</td></tr>\n",
       "<tr><td></td><td>B22</td></tr>\n",
       "<tr><td></td><td>B24</td></tr>\n",
       "<tr><td></td><td>B26</td></tr>\n",
       "<tr><td></td><td>B28</td></tr>\n",
       "<tr><td></td><td>B3</td></tr>\n",
       "<tr><td></td><td>B30</td></tr>\n",
       "<tr><td></td><td>B35</td></tr>\n",
       "<tr><td></td><td>B36</td></tr>\n",
       "<tr><td></td><td>B37</td></tr>\n",
       "<tr><td></td><td>B38</td></tr>\n",
       "<tr><td></td><td>B39</td></tr>\n",
       "<tr><td></td><td>B4</td></tr>\n",
       "<tr><td></td><td>B41</td></tr>\n",
       "<tr><td></td><td>B42</td></tr>\n",
       "<tr><td></td><td>B45</td></tr>\n",
       "<tr><td></td><td>B49</td></tr>\n",
       "<tr><td></td><td>B5</td></tr>\n",
       "<tr><td></td><td>B50</td></tr>\n",
       "<tr><td></td><td>B51 B53 B55</td></tr>\n",
       "<tr><td></td><td>B52 B54 B56</td></tr>\n",
       "<tr><td></td><td>B57 B59 B63 B66</td></tr>\n",
       "<tr><td></td><td>B58 B60</td></tr>\n",
       "<tr><td></td><td>B61</td></tr>\n",
       "<tr><td></td><td>B69</td></tr>\n",
       "<tr><td></td><td>B71</td></tr>\n",
       "<tr><td></td><td>B73</td></tr>\n",
       "<tr><td></td><td>B77</td></tr>\n",
       "<tr><td></td><td>B78</td></tr>\n",
       "<tr><td></td><td>B79</td></tr>\n",
       "<tr><td></td><td>B80</td></tr>\n",
       "<tr><td></td><td>B82 B84</td></tr>\n",
       "<tr><td></td><td>B86</td></tr>\n",
       "<tr><td></td><td>B94</td></tr>\n",
       "<tr><td></td><td>B96 B98</td></tr>\n",
       "<tr><td></td><td>C101</td></tr>\n",
       "<tr><td></td><td>C103</td></tr>\n",
       "<tr><td></td><td>C104</td></tr>\n",
       "<tr><td></td><td>C105</td></tr>\n",
       "<tr><td></td><td>C106</td></tr>\n",
       "<tr><td></td><td>C110</td></tr>\n",
       "<tr><td></td><td>C111</td></tr>\n",
       "<tr><td></td><td>C116</td></tr>\n",
       "<tr><td></td><td>C118</td></tr>\n",
       "<tr><td></td><td>C123</td></tr>\n",
       "<tr><td></td><td>C124</td></tr>\n",
       "<tr><td></td><td>C125</td></tr>\n",
       "<tr><td></td><td>C126</td></tr>\n",
       "<tr><td></td><td>C128</td></tr>\n",
       "<tr><td></td><td>C130</td></tr>\n",
       "<tr><td></td><td>C132</td></tr>\n",
       "<tr><td></td><td>C148</td></tr>\n",
       "<tr><td></td><td>C2</td></tr>\n",
       "<tr><td></td><td>C22 C26</td></tr>\n",
       "<tr><td></td><td>C23 C25 C27</td></tr>\n",
       "<tr><td></td><td>C28</td></tr>\n",
       "<tr><td></td><td>C30</td></tr>\n",
       "<tr><td></td><td>C31</td></tr>\n",
       "<tr><td></td><td>C32</td></tr>\n",
       "<tr><td></td><td>C39</td></tr>\n",
       "<tr><td></td><td>C45</td></tr>\n",
       "<tr><td></td><td>C46</td></tr>\n",
       "<tr><td></td><td>C47</td></tr>\n",
       "<tr><td></td><td>C49</td></tr>\n",
       "<tr><td></td><td>C50</td></tr>\n",
       "<tr><td></td><td>C51</td></tr>\n",
       "<tr><td></td><td>C52</td></tr>\n",
       "<tr><td></td><td>C53</td></tr>\n",
       "<tr><td></td><td>C54</td></tr>\n",
       "<tr><td></td><td>C55 C57</td></tr>\n",
       "<tr><td></td><td>C6</td></tr>\n",
       "<tr><td></td><td>C62 C64</td></tr>\n",
       "<tr><td></td><td>C65</td></tr>\n",
       "<tr><td></td><td>C68</td></tr>\n",
       "<tr><td></td><td>C7</td></tr>\n",
       "<tr><td></td><td>C70</td></tr>\n",
       "<tr><td></td><td>C78</td></tr>\n",
       "<tr><td></td><td>C80</td></tr>\n",
       "<tr><td></td><td>C82</td></tr>\n",
       "<tr><td></td><td>C83</td></tr>\n",
       "<tr><td></td><td>C85</td></tr>\n",
       "<tr><td></td><td>C86</td></tr>\n",
       "<tr><td></td><td>C87</td></tr>\n",
       "<tr><td></td><td>C89</td></tr>\n",
       "<tr><td></td><td>C90</td></tr>\n",
       "<tr><td></td><td>C91</td></tr>\n",
       "<tr><td></td><td>C92</td></tr>\n",
       "<tr><td></td><td>C93</td></tr>\n",
       "<tr><td></td><td>C95</td></tr>\n",
       "<tr><td></td><td>C97</td></tr>\n",
       "<tr><td></td><td>C99</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "<tr><td></td><td>D10 D12</td></tr>\n",
       "<tr><td></td><td>D11</td></tr>\n",
       "<tr><td></td><td>D15</td></tr>\n",
       "<tr><td></td><td>D17</td></tr>\n",
       "<tr><td></td><td>D19</td></tr>\n",
       "<tr><td></td><td>D20</td></tr>\n",
       "<tr><td></td><td>D21</td></tr>\n",
       "<tr><td></td><td>D22</td></tr>\n",
       "<tr><td></td><td>D26</td></tr>\n",
       "<tr><td></td><td>D28</td></tr>\n",
       "<tr><td></td><td>D30</td></tr>\n",
       "<tr><td></td><td>D33</td></tr>\n",
       "<tr><td></td><td>D34</td></tr>\n",
       "<tr><td></td><td>D35</td></tr>\n",
       "<tr><td></td><td>D36</td></tr>\n",
       "<tr><td></td><td>D37</td></tr>\n",
       "<tr><td></td><td>D38</td></tr>\n",
       "<tr><td></td><td>D40</td></tr>\n",
       "<tr><td></td><td>D43</td></tr>\n",
       "<tr><td></td><td>D45</td></tr>\n",
       "<tr><td></td><td>D46</td></tr>\n",
       "<tr><td></td><td>D47</td></tr>\n",
       "<tr><td></td><td>D48</td></tr>\n",
       "<tr><td></td><td>D49</td></tr>\n",
       "<tr><td></td><td>D50</td></tr>\n",
       "<tr><td></td><td>D56</td></tr>\n",
       "<tr><td></td><td>D6</td></tr>\n",
       "<tr><td></td><td>D7</td></tr>\n",
       "<tr><td></td><td>D9</td></tr>\n",
       "<tr><td></td><td>E10</td></tr>\n",
       "<tr><td></td><td>E101</td></tr>\n",
       "<tr><td></td><td>E12</td></tr>\n",
       "<tr><td></td><td>E121</td></tr>\n",
       "<tr><td></td><td>E17</td></tr>\n",
       "<tr><td></td><td>E24</td></tr>\n",
       "<tr><td></td><td>E25</td></tr>\n",
       "<tr><td></td><td>E31</td></tr>\n",
       "<tr><td></td><td>E33</td></tr>\n",
       "<tr><td></td><td>E34</td></tr>\n",
       "<tr><td></td><td>E36</td></tr>\n",
       "<tr><td></td><td>E38</td></tr>\n",
       "<tr><td></td><td>E39 E41</td></tr>\n",
       "<tr><td></td><td>E40</td></tr>\n",
       "<tr><td></td><td>E44</td></tr>\n",
       "<tr><td></td><td>E45</td></tr>\n",
       "<tr><td></td><td>E46</td></tr>\n",
       "<tr><td></td><td>E49</td></tr>\n",
       "<tr><td></td><td>E50</td></tr>\n",
       "<tr><td></td><td>E52</td></tr>\n",
       "<tr><td></td><td>E58</td></tr>\n",
       "<tr><td></td><td>E60</td></tr>\n",
       "<tr><td></td><td>E63</td></tr>\n",
       "<tr><td></td><td>E67</td></tr>\n",
       "<tr><td></td><td>E68</td></tr>\n",
       "<tr><td></td><td>E77</td></tr>\n",
       "<tr><td></td><td>E8</td></tr>\n",
       "<tr><td></td><td>F</td></tr>\n",
       "<tr><td></td><td>F E46</td></tr>\n",
       "<tr><td></td><td>F E57</td></tr>\n",
       "<tr><td></td><td>F E69</td></tr>\n",
       "<tr><td></td><td>F G63</td></tr>\n",
       "<tr><td></td><td>F G73</td></tr>\n",
       "<tr><td></td><td>F2</td></tr>\n",
       "<tr><td></td><td>F33</td></tr>\n",
       "<tr><td></td><td>F38</td></tr>\n",
       "<tr><td></td><td>F4</td></tr>\n",
       "<tr><td></td><td>G6</td></tr>\n",
       "<tr><td></td><td>T</td></tr>\n",
       "<tr><td><a name=\"embarked\">embarked</a></td><td>Cherbourg</td></tr>\n",
       "<tr><td></td><td>Queenstown</td></tr>\n",
       "<tr><td></td><td>Southampton</td></tr>\n",
       "<tr><td><a name=\"boat\">boat</a></td><td></td></tr>\n",
       "<tr><td></td><td>1</td></tr>\n",
       "<tr><td></td><td>10</td></tr>\n",
       "<tr><td></td><td>11</td></tr>\n",
       "<tr><td></td><td>12</td></tr>\n",
       "<tr><td></td><td>13</td></tr>\n",
       "<tr><td></td><td>13 15</td></tr>\n",
       "<tr><td></td><td>13 15 B</td></tr>\n",
       "<tr><td></td><td>14</td></tr>\n",
       "<tr><td></td><td>15</td></tr>\n",
       "<tr><td></td><td>15 16</td></tr>\n",
       "<tr><td></td><td>16</td></tr>\n",
       "<tr><td></td><td>2</td></tr>\n",
       "<tr><td></td><td>3</td></tr>\n",
       "<tr><td></td><td>4</td></tr>\n",
       "<tr><td></td><td>5</td></tr>\n",
       "<tr><td></td><td>5 7</td></tr>\n",
       "<tr><td></td><td>5 9</td></tr>\n",
       "<tr><td></td><td>6</td></tr>\n",
       "<tr><td></td><td>7</td></tr>\n",
       "<tr><td></td><td>8</td></tr>\n",
       "<tr><td></td><td>8 10</td></tr>\n",
       "<tr><td></td><td>9</td></tr>\n",
       "<tr><td></td><td>A</td></tr>\n",
       "<tr><td></td><td>B</td></tr>\n",
       "<tr><td></td><td>C</td></tr>\n",
       "<tr><td></td><td>C D</td></tr>\n",
       "<tr><td></td><td>D</td></tr>\n",
       "</TABLE>\n",
       "<hr>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.core.display import HTML\n",
    "HTML(filename=DATA_FOLDER+'/titanic.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each of the following questions state clearly your assumptions and discuss your findings:\n",
    "1. Describe the *type* and the *value range* of each attribute. Indicate and transform the attributes that can be `Categorical`. \n",
    "2. Plot histograms for the *travel class*, *embarkation port*, *sex* and *age* attributes. For the latter one, use *discrete decade intervals*. \n",
    "3. Calculate the proportion of passengers by *cabin floor*. Present your results in a *pie chart*.\n",
    "4. For each *travel class*, calculate the proportion of the passengers that survived. Present your results in *pie charts*.\n",
    "5. Calculate the proportion of the passengers that survived by *travel class* and *sex*. Present your results in *a single histogram*.\n",
    "6. Create 2 equally populated *age categories* and calculate survival proportions by *age category*, *travel class* and *sex*. Present your results in a `DataFrame` with unique index."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Solution\n",
    "---\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
